velox/substrait/SubstraitToVeloxPlanValidator.cpp

```diff
@@ -0,0 +1,340 @@
+    case ::substrait::Type::KindCase::kI64:
+    case ::substrait::Type::KindCase::kFp64:
+    case ::substrait::Type::KindCase::kString:
+      return true;
+    default:
+      return false;
+  }
+}
+
+bool SubstraitToVeloxPlanValidator::validateInputTypes(
+    const ::substrait::extensions::AdvancedExtension& extension,
+    std::vector<TypePtr>& types) {
+  // The input type is wrapped in enhancement.
+  if (!extension.has_enhancement()) {
+    return false;
+  }
+  const auto& enhancement = extension.enhancement();
+  ::substrait::Type inputType;
+  if (!enhancement.UnpackTo(&inputType)) {
+    return false;
+  }
+  if (!inputType.has_struct_()) {
+    return false;
+  }
+
+  // Get the input types.
+  const auto& sTypes = inputType.struct_().types();
+  for (const auto& sType : sTypes) {
+    try {
+      types.emplace_back(toVeloxType(subParser_->parseType(sType)->type));
+    } catch (const VeloxException& err) {
+      std::cout << "Type is not supported in ProjectRel due to:"
+                << err.message() << std::endl;
+      return false;
+    }
+  }
+  return true;
+}
+
+bool SubstraitToVeloxPlanValidator::validate(
+    const ::substrait::ProjectRel& sProject) {
+  if (sProject.has_input() && !validate(sProject.input())) {
+    return false;
+  }
+
+  // Get and validate the input types from extension.
+  if (!sProject.has_advanced_extension()) {
+    std::cout << "Input types are expected in ProjectRel." << std::endl;
+    return false;
+  }
+  const auto& extension = sProject.advanced_extension();
+  std::vector<TypePtr> types;
+  if (!validateInputTypes(extension, types)) {
+    std::cout << "Validation failed for input types in ProjectRel" << std::endl;
+    return false;
+  }
+
+  int32_t inputPlanNodeId = 0;
+  // Create the fake input names to be used in row type.
+  std::vector<std::string> names;
+  names.reserve(types.size());
+  for (uint32_t colIdx = 0; colIdx < types.size(); colIdx++) {
+    names.emplace_back(subParser_->makeNodeName(inputPlanNodeId, colIdx));
+  }
+  auto rowType = std::make_shared<RowType>(std::move(names), std::move(types));
+
+  // Validate the project expressions.
+  const auto& projectExprs = sProject.expressions();
+  std::vector<std::shared_ptr<const core::ITypedExpr>> expressions;
+  expressions.reserve(projectExprs.size());
+  try {
+    for (const auto& expr : projectExprs) {
+      expressions.emplace_back(exprConverter_->toVeloxExpr(expr, rowType));
+    }
+    // Try to compile the expressions. If there is any unregistred funciton or
+    // mismatched type, exception will be thrown.
+    exec::ExprSet exprSet(std::move(expressions), &execCtx_);
+  } catch (const VeloxException& err) {
+    std::cout << "Validation failed for expression in ProjectRel due to:"
+              << err.message() << std::endl;
+    return false;
+  }
+  return true;
+}
+
+bool SubstraitToVeloxPlanValidator::validate(
+    const ::substrait::FilterRel& sFilter) {
+  return false;
```

Comment:
Why does this return false unconditionally?

Reply:
We did not test this conversion in TPC-H queries, so we fallback it into spark's execution. Next, we need to test it and support the validation for it, so that the unsupported/not-verfied cases can correctly fallback.